### PR TITLE
arch: x86: Fix for PCI Multi-MSI support

### PR DESCRIPTION
**Pull request by najumon1980**
Currently Multi MSI is not working with PCI device which support
Multi MSI vector and so need to fall back to legacy mode.
This fix will enable PCI driver to use Multi MSI support
for x86 architecture.

Signed-off-by: Najumon B.A <najumon.ba@intel.com>

### CODE REVIEW

File: arch/x86/core/pcie.c [hunk 1]
```diff
@@ -214,14 +219,18 @@ uint16_t pcie_msi_mdr(unsigned int irq,
 
 	if (vector == NULL) {
 		/* edge triggered */
-		return 0x4000U | Z_IRQ_TO_INTERRUPT_VECTOR(irq);
+		msg_data_reg = 0x4000U | Z_IRQ_TO_INTERRUPT_VECTOR(irq);
 	}
-
+#if !defined(CONFIG_INTEL_VTD_ICTL)
+  else{
```
Comment (finikorg) [inline arch/x86/core/pcie.c:225]: I think space after `else` is missing, also would it look better if we put `ifdef` inside `else { }`?

File: arch/x86/core/pcie.c [hunk 2]
```diff
@@ -214,14 +219,18 @@ uint16_t pcie_msi_mdr(unsigned int irq,
...
+    msg_data_reg = vector->arch.vector;
+  }
+#endif
 	/* VT-D requires it to be 0, so let's return 0 by default */
-	return 0;
```
Comment (finikorg) [inline arch/x86/core/pcie.c:221]: The comment above does not make anymore

File: arch/x86/include/kernel_arch_func.h [hunk 3]
```diff
@@ -108,6 +108,13 @@ void z_x86_irq_connect_on_vector(unsigned int irq,
...
+ * Connect a vector for MSI/MSI-x
+ */
+void z_x86_msi_connect_on_vector(uint8_t vector,
+				 void (*func)(const void *arg),
+				 const void *arg, uint32_t flags);
+
 #endif /* !_ASMLANGUAGE */
 
-#endif /* ZEPHYR_ARCH_X86_INCLUDE_KERNEL_ARCH_FUNC_H_ */
+#endif /* ZEPHYR_ARCH_X86_INCLUDE_KERNEL_ARCH_FUNC_H_ */
```
Comment (finikorg) [inline arch/x86/include/kernel_arch_func.h:120]: This change, probably, is not needed

File: arch/x86/core/pcie.c [hunk 4]
```diff
@@ -214,14 +219,17 @@ uint16_t pcie_msi_mdr(unsigned int irq,
...
 		/* edge triggered */
-		return 0x4000U | Z_IRQ_TO_INTERRUPT_VECTOR(irq);
+		msg_data_reg = 0x4000U | Z_IRQ_TO_INTERRUPT_VECTOR(irq);
 	}
-
-	/* VT-D requires it to be 0, so let's return 0 by default */
-	return 0;
+#if !defined(CONFIG_INTEL_VTD_ICTL)
+  else {
+    msg_data_reg = vector->arch.vector;
```
Comment (jhedberg) [inline arch/x86/core/pcie.c:226]: You could avoid the #ifdef stuff completely with the help of `IS_ENABLED()`:
```
} else if (!IS_ENABLED(CONFIG_INTEL_VTD_ICTL)) {
    msg_data_reg = ...;
}
```

File: arch/x86/core/pcie.c [hunk 5]
```diff
@@ -22,6 +22,8 @@
 #ifdef CONFIG_PCIE_MMIO_CFG
 
 #define MAX_PCI_BUS_SEGMENTS 4
+#define REDDIR_HINT (8)
+#define DEST_ID_MASK (0xFF)
```
Comment (jhedberg) [inline arch/x86/core/pcie.c:26]: The `()` seem unnecessary, and also inconsistent with the previous line

File: arch/x86/core/pcie.c [hunk 6]
```diff
@@ -206,6 +210,7 @@ uint32_t pcie_msi_map(unsigned int irq,
 uint16_t pcie_msi_mdr(unsigned int irq,
 		      msi_vector_t *vector)
 {
+  uint16_t msg_data_reg = 0;
```
Comment (jhedberg) [inline arch/x86/core/pcie.c:213]: The indentation seems wrong here (spaces instead of a tab). Also, for the most part the coding style recommends an empty line after variable declarations.

**Review by jhedberg — CHANGES_REQUESTED** [review]
A few coding style comments

File: arch/x86/core/pcie.c [hunk 7]
```diff
@@ -214,14 +219,16 @@ uint16_t pcie_msi_mdr(unsigned int irq,
...
-		return 0x4000U | Z_IRQ_TO_INTERRUPT_VECTOR(irq);
+		msg_data_reg = 0x4000U | Z_IRQ_TO_INTERRUPT_VECTOR(irq);
 	}
-
-	/* VT-D requires it to be 0, so let's return 0 by default */
-	return 0;
+	else if (!IS_ENABLED(CONFIG_INTEL_VTD_ICTL) {
+		msg_data_reg = vector->arch.vector;
+	}
+#endif
```
Comment (jhedberg) [inline arch/x86/core/pcie.c:227]: Which #ifdef is this #endif matching? Did you actually try to build it?

File: arch/x86/core/pcie.c [hunk 8]
```diff
@@ -197,7 +199,9 @@ uint32_t pcie_msi_map(unsigned int irq,
 #endif
 #endif
 	{
+		uint32_t apic_id = x86_read_loapic(LOAPIC_ID) >> 24;
 		map = 0xFEE00000U; /* standard delivery to BSP local APIC */
+		map |=(((apic_id & DEST_ID_MASK) << 12) | REDDIR_HINT);
```
Comment (jhedberg) [inline arch/x86/core/pcie.c:204]: minor: missing space after `=`

File: arch/x86/core/pcie.c [hunk 9]
```diff
@@ -214,14 +219,16 @@ uint16_t pcie_msi_mdr(unsigned int irq,
 
 	if (vector == NULL) {
 		/* edge triggered */
-		return 0x4000U | Z_IRQ_TO_INTERRUPT_VECTOR(irq);
+		msg_data_reg = 0x4000U | Z_IRQ_TO_INTERRUPT_VECTOR(irq);
 	}
-
-	/* VT-D requires it to be 0, so let's return 0 by default */
-	return 0;
+	else if (!IS_ENABLED(CONFIG_INTEL_VTD_ICTL) {
```
Comment (jhedberg) [inline arch/x86/core/pcie.c:224]: This should go on the same line as the preceding `}`

File: arch/x86/core/pcie.c [hunk 10]
```diff
@@ -214,14 +219,15 @@ uint16_t pcie_msi_mdr(unsigned int irq,
 
 	if (vector == NULL) {
 		/* edge triggered */
-		return 0x4000U | Z_IRQ_TO_INTERRUPT_VECTOR(irq);
+		msg_data_reg = 0x4000U | Z_IRQ_TO_INTERRUPT_VECTOR(irq);
+	} else if (!IS_ENABLED(CONFIG_INTEL_VTD_ICTL)) {
+		msg_data_reg = vector->arch.vector;
```
Comment (andyross) [inline arch/x86/core/pcie.c:224]: This is a fix that seems to be related to VT-d, not MSI-X?   Again, can we get some clarity as to what's going on here?

**Review by andyross — CHANGES_REQUESTED** [review]
Nothing looks terribly wrong, I'm just having some trouble understanding what was wrong that this is fixing, and some of the changes seem unrelated or suprfluous.  More detail in the commit message or comments would be helpful.

File: arch/x86/core/intel64/irq.c [hunk 11]
```diff
@@ -91,6 +91,14 @@ void z_x86_irq_connect_on_vector(unsigned int irq,
 	x86_irq_args[vector - IV_IRQS] = arg;
 }
 
+void z_x86_msi_connect_on_vector(uint8_t vector,
+				 void (*func)(const void *arg),
+				 const void *arg, uint32_t flags)
+{
+	x86_irq_funcs[vector - IV_IRQS] = func;
+	x86_irq_args[vector - IV_IRQS] = arg;
```
[the remark this answers is not in the document]
Comment (najumon1980) [reply arch/x86/core/intel64/irq.c:99]: For MSI interrupt (directed from device directly to Local APIC ), IOAPIC is not involved (no need of updating RTE) neither LVT of Local APIC need to update which only for local interrupts( APIC timers etc.). So as per my understanding we should not call z_irq_controller_irq_config() for MSI/MSI-x interrupt. Also I put a TODO inside arch_pcie_msi_vectors_allocate() function for no need to allocate IRQ number as it is more relevant for device use legacy interrupt via IOAPIC or local interrupt in Local APIC.

File: arch/x86/core/pcie.c [hunk 12]
```diff
@@ -197,7 +199,9 @@ uint32_t pcie_msi_map(unsigned int irq,
 #endif
 #endif
 	{
+		uint32_t apic_id = x86_read_loapic(LOAPIC_ID) >> 24;
 		map = 0xFEE00000U; /* standard delivery to BSP local APIC */
+		map |= (((apic_id & DEST_ID_MASK) << 12) | REDIR_HINT);
```
[the remark this answers is not in the document]
Comment (najumon1980) [reply arch/x86/core/pcie.c:204]: Ideally enable_msix() PCI host driver's API should provide support for user/device driver to specify Interrupt Affinity since MSI support affinity per device and MSI-X provide per vector for Intel SOCs (via MSI data register).  If we able to update PCI Host interface with this should be great and I can provide another API says "pcie_msi_set_cpu_affinity()" and can assign these flag accordingly. Please share your suggestion ?

File: arch/x86/core/pcie.c (hunk 10 above)
Comment (najumon1980) [reply arch/x86/core/pcie.c:224]: This added to keep previous logic for VT-d and legacy interrupt intact . Also for MSI/MSI-X the irq number is not relevant (please see my previous comment) and  hence for MSI we cannot use Z_IRQ_TO_INTERRUPT_VECTOR() macro. 
But I believe it is better remove the MSI-X check so that the API honor both MSI and MSI-X equally as show bellow:
uint16_t pcie_msi_mdr(unsigned int irq,
		      msi_vector_t *vector)
{
	uint16_t msg_data_reg = 0;

	if ((vector != NULL) && (IS_ENABLED(CONFIG_PCIE_MSI) || 
			IS_ENABLED(CONFIG_PCIE_MSI_X) ) {
		msg_data_reg  = vector->arch.vector;
	}
	else if (vector == NULL)) {
		/* edge triggered */
		msg_data_reg  = 0x4000U | Z_IRQ_TO_INTERRUPT_VECTOR(irq);
	}

	/* VT-D requires it to be 0, so let's return 0 by default */
	return msg_data_reg;
}

**Comment by najumon1980** [comment]
@andyross , could you please respond to my comment ?

**Comment by andyross** [comment]
Sorry, I guess I wasn't clear: I wasn't asking those question so you could answer them in github comments, I was saying you need answers for that stuff in the patches themselves so future developers understand what's happening. ;)

File: arch/x86/core/intel64/irq.c (hunk 11 above)
[the remark this answers is not in the document]
Comment (andyross) [reply arch/x86/core/intel64/irq.c:99]: You can't put an "is this an MSI interrupt?" test in the existing connect function?

File: arch/x86/core/pcie.c (hunk 12 above)
[the remark this answers is not in the document]
Comment (andyross) [reply arch/x86/core/pcie.c:204]: Right, we don't have a decent solution for targetting MSI interrupts currently, we need to do some design work.

I was just asking why you were changing this.  Is it needed for the fix?  Why?

File: arch/x86/core/intel64/irq.c (hunk 11 above)
[the remark this answers is not in the document]
Comment (najumon1980) [reply arch/x86/core/intel64/irq.c:99]: My view it is better have a separate API for MSI vs Legacy

File: arch/x86/core/pcie.c (hunk 12 above)
[the remark this answers is not in the document]
Comment (najumon1980) [reply arch/x86/core/pcie.c]: I will create another commit for adding new MSI API for setting affinity and hence removed the changes

**Comment by najumon1980** [comment]
@andyross, updated/responded to your review comments and also put the commit message with few more details.   Please let me know any additional info/changes required and if not please help to merge the patch.

File: arch/x86/core/pcie.c [hunk 13]
```diff
@@ -18,11 +18,11 @@
 #include <drivers/pcie/msi.h>
 #endif
 
+#define MAX_PCI_BUS_SEGMENTS 4
```
Comment (tbursztyka) [inline arch/x86/core/pcie.c:21]: Why this change?

File: arch/x86/core/pcie.c [hunk 14]
```diff
@@ -206,38 +206,19 @@ uint32_t pcie_msi_map(unsigned int irq,
...
-#endif
+	uint16_t msg_data_reg = 0;
 
-	if (vector == NULL) {
+	if (vector != NULL) {
+		msg_data_reg = vector->arch.vector;
+	} else {
 		/* edge triggered */
-		return 0x4000U | Z_IRQ_TO_INTERRUPT_VECTOR(irq);
+		msg_data_reg = 0x4000U | Z_IRQ_TO_INTERRUPT_VECTOR(irq);
```
Comment (tbursztyka) [inline arch/x86/core/pcie.c:215]: that is not going to work if there is no default irq set for the msi device, see below

File: arch/x86/core/intel64/irq.c (hunk 11 above)
[the remark this answers is not in the document]
Comment (tbursztyka) [reply arch/x86/core/intel64/irq.c:99]: @andyross We don't store anywhere the trigger type of a vector (irq through IOAPIC or MSI one). So at this stage it is not possible to check for this.

That said, having z_x86_msi_connect_on_vector() is not a good idea. Better have a z_x86_connect_on_vector().
Which one would also be used by the above z_x86_irq_connect_on_vector().

File: arch/x86/core/pcie.c [hunk 15]
```diff
@@ -270,20 +251,23 @@ uint8_t arch_pcie_msi_vectors_allocate(unsigned int priority,
 #endif /* CONFIG_INTEL_VTD_ICTL */
 
 		for (i = 0; i < n_vector; i++) {
-			uint32_t data;
-
-			data = _read_pcie_irq_data(vectors[i].bdf);
-
-			vectors[i].arch.irq = pcie_alloc_irq(vectors[i].bdf);
```
Comment (tbursztyka) [inline arch/x86/core/pcie.c:277]: This is indeed useless for MSI as it does not make any use of IRQ.
However, due to a design issue on how zephyr manages x86 interrupt vector and irq, you cannot remove that or else line 215 might fail (single msi device) badly if no irq has been allocated to the device(s): irq will be 0.

Zephyr stores vectors on irq basis. It's bad, it should store vector and irq if relevant, in a dedicated data structure. I knew this problem from the start when starting to work on vt-d but no time to change it all (It's quite a big rework actually), so I kept this behavior: it's not generating any harm since as soon as a device is meant to use MSI it will never use the allocated IRQ thus what's being done on IOAPIC will remain dead silent.

File: arch/x86/core/pcie.c [hunk 16]
```diff
@@ -270,20 +251,23 @@ uint8_t arch_pcie_msi_vectors_allocate(unsigned int priority,
...
 			if (vectors[i].arch.vector < 0) {
 				return 0;
 			}
 
+#ifdef CONFIG_PCIE_MSI_X
+			if (!vectors[0].msix)
+#endif
+			{
+				if (prev_vector != -1) {
+					__ASSERT(prev_vector == (vectors[i].arch.vector - 1),
```
Comment (tbursztyka) [inline arch/x86/core/pcie.c:266]: This is exactly why VT-D was selected for generating multi-vector MSI. Zephyr's vector allocator is not bullet proof on contiguous allocation. And you really cannot fail here like that since this will happen very often. Let's take ehl's ethernet stuff: they require 32 vectors (on which only 16 are actually used, 15 at the beginning and the last one at the end. Don't ask me why they went for this weird setup, in addition of using msi instead of msi-x...), and there can be 2 device instances, and more devices may require such feature (who knows what devices ppl will plug in). Depending on priority and initialization order, the vector array is soon going to be like an emmental cheese: too many holes and no places to fit such long contiguous vectors request.

Note that this is the same issue on Linux: it cannot allocate contiguous vectors and make use of VT-D for MSI multi-vector purpose.

**Review by tbursztyka — CHANGES_REQUESTED** [review]
That's not going to fly. See my comments.

There are 2 issues here: 
- Zephyr does a bad job at managing irq/vectors for x86 and what you are trying to fix for MSI (so it does not also allocate an RTE on IOAPIC side) is not going to work.
- contiguous allocation of vectors is not going to be bullet proof due to the allocation logic in zephyr (about the same as in Linux), thus why vt-d work was requested.

File: arch/x86/core/pcie.c (hunk 15 above)
Comment (najumon1980) [reply arch/x86/core/pcie.c:277]: We should request x86 porting team to fix this issue rather than doing this work around. I can put back this code until Zephyr fix this issue.

File: arch/x86/core/pcie.c (hunk 16 above)
Comment (najumon1980) [reply arch/x86/core/pcie.c:266]: This is the limitation of MSI by design and is addressed by introducing MSI-X in later PCI spec. So for device exposing MSI capability, system SW have two choice either allocate vector as power of two requested by MSI cap register or emulate as MSI-X. We have to keep this 1st option currently (which is also by design correct only) and we don't have 2nd working option currently available.

**Comment by najumon1980** [comment]
Regarding VT-d: As per my opinion VT-d is not must to have and only nice to have which address the limitation of MSI such as need contiguous vector by emulating MSI device as MSI-x. 

Regarding the existing unnecessary IOAPIC programming/allocating IRQ WA code: For MSI capable device IOAPIC will not come into picture at all and is only require for legacy IPs which are not have MSI generators. So we should request to fix Zephyr x86 interrupt design rather than doing this WA. I will put back the existing WA code since it should not harm until fixed the Zephyr issue (once MSI interrupt enabled,  device will not generate sideband interrupt).

My option is to keep this change (after put back the WA) and merge if you not see any issue.  VT-d will be additional option for user to use vector more efficiently

File: arch/x86/core/pcie.c (hunk 15 above)
Comment (najumon1980) [reply arch/x86/core/pcie.c:277]: line 215 will not have any impact on removing this code since this code will come into picture only for multi vector (n_vector > 1). But any way will put back this code with TODO comment to fix the design

File: arch/x86/core/pcie.c (hunk 14 above)
Comment (najumon1980) [reply arch/x86/core/pcie.c:215]: This will not have any impact on removing code from line number 273 to 278, since that code will come into picture only for multi vector (n_vector > 1). But any way will put back that code with TODO comment to fix the design

File: arch/x86/core/pcie.c (hunk 15 above)
Comment (finikorg) [reply arch/x86/core/pcie.c:277]: For 32 vectors this code makes lots of unneeded PCI requests, I was also proposing removing this code.

**Review by finikorg — APPROVED** [review]
The code might be improved but I think it makes sense to merge it since it fixes MSI support, broken at the moment.

**Comment by tbursztyka** [comment]
@finikorg MSI is not broken, multi-vector MSI never worked (only single MSI). It's a missing feature.

@najumon1980 vt-d is not an option. It's the only way to go for an always working multi-vector MSI support. Your patch is a hack that will not survive production. (And VT-D has nothing to do with emulating MSI into MSI-x).

**Comment by najumon1980** [comment]
@tbursztyka, as per my understanding, VT-d is designed for improve device performance on virtualized environment and not for address multi MSI issue (though it can avoid contiguous vector need). If system SW not enabled  VT-d for interrupt remapping (IRES is not set) or Interrupt format bit not set in MSI address register then still the MSI interrupt work in Compatibility format. Also multi MSI is a feature supported even before VT-d introduced and should expect to work if platform not enabled VT-d and what this PR try to enable multi MSI without VT-d.

**So I believe, Vt-d is not must to have and is nice to have for efficient use of interrupt vectors .  This patch is as per MSI mode specified by Intel specification and is not a hack.**

**Comment by najumon1980** [comment]
I had created following PR for VT-d fix and with that Multi MSI seams working. 
**https://github.com/zephyrproject-rtos/zephyr/pull/36725**

Also as per my opinion, we should support Multi MSI without VT-d as well and hence  I will create multiple commitment for the new VT-d PR including this changes

**@jhedberg please share your suggestion ?**

**Comment by jhedberg** [comment]
@najumon1980 there are some compliance issue like this one:
```
-:104: ERROR:SPACING: space prohibited before that '++' (ctx:WxO)
#104: FILE: arch/x86/core/pcie.c:254:
+			priority ++;
```

File: arch/x86/core/pcie.c [hunk 17]
```diff
@@ -206,22 +206,19 @@ uint32_t pcie_msi_map(unsigned int irq,
...
+	if (vector != NULL) {
+		msg_data_reg = vector->arch.vector;
+	} else {
 		/* edge triggered */
-		return 0x4000U | Z_IRQ_TO_INTERRUPT_VECTOR(irq);
+		msg_data_reg = 0x4000U | Z_IRQ_TO_INTERRUPT_VECTOR(irq);
 	}
-
-	/* VT-D requires it to be 0, so let's return 0 by default */
-	return 0;
```
Comment (michael1schmidt) [inline arch/x86/core/pcie.c:221]: this is the default code branch executed if neither MSI-X nor Multi-MSI capabilities are available and it is assumed that this branch is used only if VT-D is enabled. But it is also executed when only single vector MSI should be used. This would lead in a failure as the MSI vector used to trigger an IRQ event will be wrong.

File: arch/x86/core/pcie.c [hunk 18]
```diff
@@ -206,22 +206,19 @@ uint32_t pcie_msi_map(unsigned int irq,
...
+#endif
+	return msg_data_reg;
 }
 
-#if defined(CONFIG_INTEL_VTD_ICTL) || defined(CONFIG_PCIE_MSI_X)
```
Comment (michael1schmidt) [inline arch/x86/core/pcie.c:224]: below code should be executed also when CONFIG_MSI is set in order to make the IRW/MSI related functions available when using MSI interrupt signaling.

File: arch/x86/core/pcie.c [hunk 19]
```diff
@@ -236,6 +236,74 @@ static inline void _write_pcie_irq_data(pcie_bdf_t bdf, uint32_t data)
...
+						uint8_t n_vector)
+{
+	const int MAX_PRIORITY = 13;
+	int vector = 0, prev_vector = -1, start_vector;
+	int i;
+
+	while (vector != -1) {
+		vector = z_x86_allocate_vector(priority, prev_vector);
+		if ((vector == -1) && (priority <= MAX_PRIORITY)) {
+#if defined (CONFIG_MULTI_MSI_ANY_PRIORITY)
```
Comment (tbursztyka) [inline arch/x86/core/pcie.c:251]: to declutter things when there is a lot of ifdefs everywhere. Use if (IS_ENABLED()) {} else {} whenever possible (when no function/variable are out of the definition)

File: arch/x86/core/pcie.c [hunk 20]
```diff
@@ -236,6 +236,74 @@ static inline void _write_pcie_irq_data(pcie_bdf_t bdf, uint32_t data)
...
+#endif
+		} else if (priority >= MAX_PRIORITY) {
+			break;
+		}
+
+		/* n_vector assume always power of 2 */
+		if (!(vector % n_vector)) {
+			start_vector = vector;
+		} else {
+#if defined (CONFIG_MULTI_MSI_ANY_PRIORITY)
```
Comment (tbursztyka) [inline arch/x86/core/pcie.c:269]: same here

File: arch/x86/core/pcie.c [hunk 21]
```diff
@@ -236,6 +236,74 @@ static inline void _write_pcie_irq_data(pcie_bdf_t bdf, uint32_t data)
...
+				break;
+			}
+			prev_vector = vector;
+			vectors[i].arch.vector = vector;
+		}
+
+		if (i == n_vector) {
+			return n_vector;
+		} else {
+#if defined (CONFIG_MULTI_MSI_ANY_PRIORITY)
```
Comment (tbursztyka) [inline arch/x86/core/pcie.c:295]: same here

File: arch/x86/core/pcie.c (hunk 16 above)
Comment (tbursztyka) [reply arch/x86/core/pcie.c]: Remove that change altogether, since next commit is going to fix it properly.

File: arch/x86/core/pcie.c [hunk 22]
```diff
@@ -308,11 +317,10 @@ bool arch_pcie_msi_vector_connect(msi_vector_t *vector,
 	}
 #endif /* CONFIG_INTEL_VTD_ICTL */
 
-	z_x86_irq_connect_on_vector(vector->arch.irq, vector->arch.vector,
-				    routine, parameter, flags);
+	z_x86_connect_on_vector(vector->arch.vector,
+				    routine, parameter);
```
Comment (tbursztyka) [inline arch/x86/core/pcie.c:386]: re-align the indentation

File: arch/x86/core/pcie.c [hunk 23]
```diff
@@ -206,22 +206,19 @@ uint32_t pcie_msi_map(unsigned int irq,
...
+	} else {
 		/* edge triggered */
-		return 0x4000U | Z_IRQ_TO_INTERRUPT_VECTOR(irq);
+		msg_data_reg = 0x4000U | Z_IRQ_TO_INTERRUPT_VECTOR(irq);
 	}
-
-	/* VT-D requires it to be 0, so let's return 0 by default */
-	return 0;
+#endif
+	return msg_data_reg;
```
Comment (tbursztyka) [inline arch/x86/core/pcie.c:219]: all the changes here in pcie_msi_mdr() do not fit in this commit anymore, please move it into the next one.

File: arch/x86/core/intel64/irq.c [hunk 24]
```diff
@@ -80,15 +80,22 @@ int z_x86_allocate_vector(unsigned int priority, int prev_vector)
 	return -1;
```
Comment (tbursztyka) [inline arch/x86/core/intel64/irq.c:80]: update the commit message accordingly

**Review by tbursztyka — CHANGES_REQUESTED** [review]

**Comment by jhedberg** [comment]
What's the story with this PR? Is the stale flag appropriate? @najumon1980 are you still working on this?

**Comment by najumon1980** [comment]
hi Johan, this fix is required and I will update with the review comments
